### DISCUSSION TITLE
Creating basic hello world for OnAIR linking with cFS

### DISCUSSION THREAD

**Discussion by aerospaceadversarylab** [nasa/cFS#801] [General]
We have been using [this integrated repository](https://github.com/the-other-james/cFS/tree/OnAIR-integration) which uses sample_app with OnAIR to run alongside cFS. What if we want to make some basic Python script, like a hello world application, and have it be displayed in either OnAIR's `driver.py` or cFS's logs. Can we get some guidance on how to do this? This is a precursor of having our own model running in OnAIR. Our efforts thus far have gotten us to run a `hello_world.ini` file which attempted to try to link a `hello_world.py` file under `[MODEL]`, but when the simulation starts nothing happens?
![image](https://github.com/user-attachments/assets/a3702c4f-4fdb-4c26-ba18-7a583cb936ad)
![image](https://github.com/user-attachments/assets/a77b17c0-25e1-4ba6-82cd-081daba7cfec)
![image](https://github.com/user-attachments/assets/2a04fb9b-41e6-41c5-98c3-1859bae59680)

Any advice would be appreciated!

**Comment by skliper** [comment]
Ping @the-other-james @williamzhang0306

**Reply by aerospaceadversarylab** [reply]
Following up on this. We want to try running the Kalman filtering plugin that's in the source code of OnAIR. But when we do we keep getting path errors. 
![image](https://github.com/user-attachments/assets/a2ebf047-baab-491c-bbd4-750e3a96bd7f)
It seems that `driver.py` is not expanding the relative file paths properly. When I tried to put in an absolute path, I got another error because it seems the path string being used is still relative.

**Reply by williamzhang0306** [reply]
Hi, I don't maintain that repo or work on OnAIR anymore but I'd be happy to help out.

The pathing with OnAIR is a little weird, but if I recall correctly the plugin path provided in the `.ini` file should be relative to the working directory when driver.py is ran. The Kalman filter example provided in the OnAIR source code is assuming that onair is the working directory. 

I suspect that changing these two following lines in `cf/onair/onair/config/kalman_csv_output_example.ini` might fix the kalman filter plugin not being found.

from:
```
KnowledgeRepPluginDict = {'Kalman Filter': 'plugins/kalman'}
LearnersPluginDict = {'csv output':'plugins/csv_output'}
```

to
```
KnowledgeRepPluginDict = {'Kalman Filter': 'cf/onair/plugins/kalman'}
LearnersPluginDict = {'csv output':'cf/onair/plugins/csv_output'}
```

Let me know if that helps!

**Reply by aerospaceadversarylab** [reply]
```
root@5e1e8b395751:/cFS-fsw/build/exe/cpu1# cat cf/onair/onair/config/kalman_csv_output_example.ini 
[DEFAULT]
TelemetryDataFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
TelemetryFile = 700_crash_to_earth_1.csv
TelemetryMetadataFilePath = onair/data/telemetry_configs/
MetaFile = data_physics_generation_CONFIG.json
ParserFileName = onair/data_handling/csv_parser.py

KnowledgeRepPluginDict = {'Kalman Filter': 'cf/onair/plugins/kalman'}
LearnersPluginDict = {'csv output':'cf/onair/plugins/csv_output'}
PlannersPluginDict = {}
ComplexPluginDict = {}

[RUN_FLAGS]
IO_Flag = true
root@5e1e8b395751:/cFS-fsw/build/exe/cpu1# python3 cf/onair/driver.py cf/onair/onair/config/kalman_csv_output_example.ini
Traceback (most recent call last):
  File "cf/onair/driver.py", line 133, in <module>
    main()
  File "cf/onair/driver.py", line 103, in main
    OnAIR = ExecutionEngine(args.configfile, save_name, args.save)
  File "/cFS-fsw/build/exe/cpu1/cf/onair/onair/src/run_scripts/execution_engine.py", line 59, in __init__
    self.parse_data(self.parser_file_name, self.fullTelemetryFileName, self.fullMetaDataFileName)
  File "/cFS-fsw/build/exe/cpu1/cf/onair/onair/src/run_scripts/execution_engine.py", line 110, in parse_data
    data_source_spec.loader.exec_module(data_source_module)
  File "<frozen importlib._bootstrap_external>", line 844, in exec_module
  File "<frozen importlib._bootstrap_external>", line 980, in get_code
  File "<frozen importlib._bootstrap_external>", line 1037, in get_data
FileNotFoundError: [Errno 2] No such file or directory: '/cFS-fsw/build/exe/cpu1/onair/data_handling/csv_parser.py'
```

**Reply by the-other-james** [reply]
Where is csv_parser.py located? Right now OnAIR is looking in '/cFS-fsw/build/exe/cpu1/onair/data_handling/' as per the error message... but I'm guessing that should actually be '/cFS-fsw/build/exe/cpu1/cf/onair/data_handling/' or '/cFS-fsw/build/exe/cpu1/cf/onair/onair/data_handling/'

Have you tried updating your .ini file to set ParserFileName to 'cf/onair/data_handling/csv_parser.py' or 'cf/onair/onair/data_handling/csv_parser.py'?

**Reply by aerospaceadversarylab** [reply]
Ok that worked, but I just realized that the `TelemetryDataFilePath`: `cf/onair/onair/data/raw_telemetry_data/data_physics_generation/Errors` is not populated with any csv files. This is in the `build/exe/cpu1` directory. But the original directory before being built: `/cFS-fsw/apps/onair_app/onair/onair/data/raw_telemetry_data/data_physics_generation/Errors` has all the correct csv files we need, so when we go to run the `driver.py` we get a `FileNotFound` error.

**Reply by aerospaceadversarylab** [reply]
After copying the csv file to the correct directory we got the Kalman plugin to work. Do you know what the expected output is? We're getting this as our output which seems correct but we can't verify it.
![image](https://github.com/user-attachments/assets/f52ffad3-a45c-410a-8670-375d6ff71ab5)
